src/frontend/src/optimizer/plan_node/logical_join.rs

```diff
@@ -360,19 +359,24 @@ impl LogicalJoin {
         let order_key = table_desc.order_column_indices();
         let dist_key = table_desc.distribution_key.clone();
         // The at least prefix of order key that contains distribution key.
-        let at_least_prefix_len = {
-            let mut max_pos = 0;
-            for d in dist_key {
-                max_pos = max(
-                    max_pos,
-                    order_key
-                        .iter()
-                        .position(|&x| x == d)
-                        .expect("dist_key must in order_key"),
-                );
-            }
-            max_pos + 1
-        };
+        let mut dist_key_in_order_key_pos = vec![];
+        for d in dist_key {
+            let pos = order_key
+                .iter()
+                .position(|&x| x == d)
+                .expect("dist_key must in order_key");
+            dist_key_in_order_key_pos.push(pos);
+        }
+        // The at least prefix of order key that contains distribution key.
+        let at_least_prefix_len = dist_key_in_order_key_pos
+            .iter()
+            .max()
+            .map_or(0, |pos| pos + 1);
+
+        // Distributed lookup join can't support lookup table with a singleton distribution.
+        if at_least_prefix_len == 0 {
```

Comment:
For the distributed join, using a simple singleton distribution is not enough because we need to schedule the distributed lookup join executor together with its lookup table (whose distribution must be a singleton in this case). That means these two singleton distributions should have a connection and we need to add a `TableId` into the singleton distribution just like ` UpstreamHashShard(Vec<usize>, TableId)`. Maybe we can support it in the future.

Reply:
Yes, I know what you mean. This `SingletonExchange` needs to shuffle the left table scan records to the right table. Our current singleton distribution doesn't have this information. On the other hand, if the left table scan is also a singleton, then this exchange could be removed which is unexpected since the `DistributedLookupJoin` needs to schedule with the right table rather than the left table.